### PR TITLE
(feat) basic rename

### PR DESCRIPTION
**Pull request by dummdidumm**
#110

- [x] rename where everything is inside same svelte file works
- [x] rename of variable inside svelte which is also in another ts/js file works
- [ ] rename of svelte prop works when doing the rename for prop of component A in component B which uses that prop
- [x] rename of svelte prop works when doing the rename for prop of component A in component A

Out of scope for this PR:
-  rename in ts file which would affect svelte file -> out of this scope because we would also need to trigger the language server on ts/js files then

### CODE REVIEW

**Comment by jasonlyu123** [comment]
Found an issue. if I rename a props that is html attribute like disabled. It would do a rename on the html attribute no matter that props exist on the component or not.

**Comment by dummdidumm** [comment]
Good catch, thank you. There is also a prop rename problem. I think I need to enhance `svelte2tsx` with some diagnostic capabilities like "what other svelte components does this file import?" or "what props does this file have?".

**Comment by dummdidumm** [comment]
I ended up using `getRenameInfo` of the ts language service. The downside is that component properties can only be renamed within that component now. I added a TODO about this, this can be tackled later. I also added `prepareRename` support so it shows the user early that this cannot be renamed.

**Comment by dummdidumm** [comment]
@jasonlyu123 could you take one more look? 😃 Also code-wise, I had to jump through some hoops to get it working.

File: packages/language-server/src/lib/documents/utils.ts [hunk 1]
```diff
@@ -215,3 +215,33 @@ export function isInTag(position: Position, tagInfo: TagInformation | null): boo
 export function getTextInRange(range: Range, text: string) {
     return text.substring(offsetAt(range.start, text), offsetAt(range.end, text));
 }
+
+export function getVariableAtPosition(position: Position, text: string) {
+    return getVariableAtOffset(offsetAt(position, text), text);
+}
+
+export function getVariableAtOffset(offset: number, text: string) {
```
Comment (jasonlyu123) [inline packages/language-server/src/lib/documents/utils.ts:223]: also variable.b, variable) and operators without space in between.

**Comment by jasonlyu123** [comment]
```svelte
<input bind:value />
```
would be renamed to
```svelte
<input bind:valuvalues />
```

File: packages/language-server/src/plugins/typescript/features/RenameProvider.ts [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        fragment: SvelteSnapshotFragment,
+        position: Position,
+        convertedRenameLocations: (ts.RenameLocation & { range: Range })[],
+        fragments: Map<string, SnapshotFragment>,
+        lang: ts.LanguageService,
+    ) {
+        // First find out if it's really the "rename prop inside component with that prop" case
+        // Use original document for that because only there the `export` is present.
+        const regex = new RegExp(
+            `export\\s+(const|let)\\s+${getVariableAtPosition(
```
[the remark this answers is not in the document]
Comment (jasonlyu123) [reply packages/language-server/src/plugins/typescript/features/RenameProvider.ts:152]: @sw-yx do you know about export const, function, class? I only found it to be available in the component instance, if so it probably doesn't needs to be checked.

File: packages/language-server/src/lib/documents/utils.ts (hunk 1 above)
Comment (jasonlyu123) [reply packages/language-server/src/lib/documents/utils.ts:223]: I found a hack that can be used to find the identifier/symbol in the generated tsx:
```ts
const offset = fragment.offsetAt(fragment.getGeneratedPosition(position));
const { start, length } = lang.getSmartSelectionRange(tsDoc.filePath, offset).textSpan;
const identifier = tsDoc.getText(start, start + length);
```

**Comment by dummdidumm** [comment]
This is a source map problem, seems like `svelte2tsx` mappings are off after conversion. I was able to fix it. I also added your suggestion for the variable retrieval and added a TODO that `export {x as y}` syntax is currently not supported for rename. The function/class thing ... let's just wait till someone shows up with this request 😄

**Pull request merged**